### PR TITLE
fix(soql): disable remote query checks. Make it an experimental option

### PR DESCRIPTION
**Pull request by dobladez**
### What issues does this PR fix or reference?
@W-8611581@

### Functionality Before
SOQL editor would always call the remote API to validate queries

### Functionality After
This "experimental" feature is now disabled by default. Added new configuration option for the user to enable it.

https://user-images.githubusercontent.com/152839/103817675-e501d480-5045-11eb-8f54-36ac25cec8b3.mov

### CODE REVIEW

File: packages/salesforcedx-vscode-soql/package.nls.json [hunk 1]
```diff
@@ -1,4 +1,6 @@
 {
+  "configuration_title": "Salesforce SOQL Configuration",
   "soqlCustom_soql": "SOQL Builder",
-  "soql_builder_open_new": "SFDX: Create Query in SOQL Builder"
+  "soql_builder_open_new": "SFDX: Create Query in SOQL Builder",
+  "soql_editor_remote_checks_flag": "Validate SOQL editor queries against your Org. When enabled, the SOQL editor calls the remote API to validate the query on your default Org environment and provide better syntax checks."
```
Comment (ekapner) [inline packages/salesforcedx-vscode-soql/package.nls.json:5]:
```suggestion
  "soql_editor_remote_checks_flag": "Validate SOQL editor queries against your org. When enabled, the SOQL editor calls the remote API to validate the query on your default org and provide better syntax checks."
```

Comment (ekapner) [inline packages/salesforcedx-vscode-soql/package.nls.json:5]: @praksb @dobladez - do we need to say SOQL "editor" queries - or can we just say "SOQL queries"? Is this to distinguish a query built in a text editor vs. SOQL Builder?

Comment (ekapner) [reply packages/salesforcedx-vscode-soql/package.nls.json:5]: Also, couldn't figure out how to put my comment closer to the .mov. Can we all cap SOQL "Experimental: SOQL Editor Remote Checks" - but still want to make sure I understand the context for 'editor' here as well

Comment (dobladez) [reply packages/salesforcedx-vscode-soql/package.nls.json:5]: Good point. Yes, this flag only affects the behavior of the text editor... Should we use the full "SOQL text editor queries" ?

File: packages/salesforcedx-vscode-soql/package.json [hunk 2]
```diff
@@ -136,6 +136,17 @@
           "when": "sfdx:project_opened"
         }
       ]
+    },
+    "configuration": {
+      "type": "object",
+      "title": "%configuration_title%",
+      "properties": {
+        "salesforcedx-vscode-soql.experimental.soqlEditorRemoteChecks": {
```
Comment (praksb) [inline packages/salesforcedx-vscode-soql/package.json:144]: can we just call it soqlValidation instead of soqlEditorRemoteChecks?

File: packages/salesforcedx-vscode-soql/package.json [hunk 3]
```diff
@@ -136,6 +136,17 @@
...
+          "type": "boolean",
+          "default": false,
+          "description": "%soql_editor_remote_checks_flag%"
```
Comment (praksb) [inline packages/salesforcedx-vscode-soql/package.json:147]: soql_validation_flag instead?

File: packages/salesforcedx-vscode-soql/package.nls.json (hunk 1 above)
Comment (praksb) [reply packages/salesforcedx-vscode-soql/package.nls.json:5]: Even though its applicable only to queries in the text editor, I don't think that distinction needs to be explicitly called out in the description text.

Comment (praksb) [reply packages/salesforcedx-vscode-soql/package.nls.json:5]: I would make this description a little more compact. Maybe something like "Enable SOQL query validation using the REST API and display errors".

File: packages/salesforcedx-vscode-soql/test/vscode-integration/client/client.test.ts [hunk 4]
```diff
@@ -69,16 +61,57 @@ describe('SOQL language client', () => {
...
+  });
+
+  it('should not create diagnostics based off of remote query validation when disabled', async () => {
+    soqlFileUri = await writeSOQLFile(
+      'testSemanticErrors_remoteRunDefault.soql',
+      'SELECT Ids FROM Account'
+    );
+
+    stubSOQLExtensionConfiguration(sandbox, {
+      'experimental.soqlEditorRemoteChecks': true
```
Comment (praksb) [inline packages/salesforcedx-vscode-soql/test/vscode-integration/client/client.test.ts:89]: I think this test needs to be fixed.  'experimental.soqlEditorRemoteChecks': true and the test is passing.

1) The test should be fixed so that the mock connection throws an error
2) experimental.soqlEditorRemoteChecks should be set to false

**Review by praksb — CHANGES_REQUESTED** [review]
I just had a few small comments on the text description.

Also, Can you take a look at my comment on one of the tests?

File: packages/salesforcedx-vscode-soql/package.nls.json (hunk 1 above)
Comment (ekapner) [reply packages/salesforcedx-vscode-soql/package.nls.json:5]: @praksb - If I'm understanding your point, are you saying that we should just use "SOQL queries"?

File: packages/salesforcedx-vscode-soql/test/vscode-integration/client/client.test.ts (hunk 4 above)
Comment (dobladez) [reply packages/salesforcedx-vscode-soql/test/vscode-integration/client/client.test.ts]: Great catch!!!  Fixed now. Very tricky:

Seems like the integration test framework runs a failed test multiple times.  And in the case of language diagnostics, the second time it run for the exact same file and input, no diagnostics check happens (I guess this is an optimization on VSCode. That's also how it haves when testing manually).

Here's what happened:
1. Assertion `expect(querySpy.notCalled).to.be.true` failed
2. Test re-run
3. VSCode didn't run the diagnostic check again (optimization)
4. Thus, assertion `expect(querySpy.notCalled).to.be.true` and the rest passed because no validation happened

Thanks!

File: packages/salesforcedx-vscode-soql/package.nls.json (hunk 1 above)
Comment (praksb) [reply packages/salesforcedx-vscode-soql/package.nls.json:5]: @ekapner yes. i am suggesting using just soql queries.

**Review by praksb — COMMENTED** [review]
Test fix looks good. Let me know what you think about the changes to the description text. Then we can move this to review and try to push it through by Monday.

File: packages/salesforcedx-vscode-soql/package.nls.json (hunk 1 above)
Comment (ekapner) [reply packages/salesforcedx-vscode-soql/package.nls.json:5]: I don't think we need to mention better syntax checks. Is the fact that it's REST API important? Validate SOQL queries and identify errors. (?)

Comment (praksb) [reply packages/salesforcedx-vscode-soql/package.nls.json:5]: The reason I want to put "REST API" is that it will tell the customers that the checks happens on the server side. That way they would (hopefully) know that this validation wouldn't work if they are offline or having an auth issue and/or not connected to an org.

**Comment by dobladez** [comment]
@praksb  @ekapner Is the text below what we agreed on? Thanks!

image

**Comment by dobladez** [comment]
Final ;-)

image

File: .vscode/launch.json [hunk 5]
```diff
@@ -595,29 +595,6 @@
...
-      "smartStep": true,
-      "outFiles": ["${workspaceFolder}/packages/*/out/**/*.js"],
-      "sourceMapPathOverrides": {
-        "../../../test/vscode-integration": "${workspaceFolder}/packages/salesforcedx-vscode-soql/out/test/vscode-integration/*",
-        "../../../../test/vscode-integration": "${workspaceFolder}/packages/salesforcedx-vscode-soql/out/test/vscode-integration/*",
-        "webpack:///salesforcedx-utils-vscode/./*": "${workspaceFolder}/packages/salesforcedx-utils-vscode/*",
-        "webpack:///*": "*"
-      },
-      "preLaunchTask": "Compile",
-      "internalConsoleOptions": "openOnSessionStart"
```
Comment (praksb) [inline .vscode/launch.json:620]: Did you remove this by mistake?

File: packages/salesforcedx-vscode-soql/src/client/client.ts [hunk 6]
```diff
@@ -60,28 +53,7 @@ export async function startLanguageClient(
...
-        const items = (await next(
-          document,
-          position,
-          context,
-          token
-        )) as ProtocolCompletionItem[];
-
-        return expandPlaceholders(items);
-      }
-    }
```
Comment (praksb) [inline packages/salesforcedx-vscode-soql/src/client/client.ts:84]: this shouldn't be part of this PR right?

File: packages/salesforcedx-vscode-soql/src/client/codeCompletion.ts [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        return fieldItems;
+      }, [] as ProtocolCompletionItem[]);
+      return sobjectFields;
+    } catch (metadataError) {
+      return [];
+    }
+  }
+};
+
+function newFieldCompletionItem(
```
Comment (praksb) [inline packages/salesforcedx-vscode-soql/src/client/codeCompletion.ts:102]: I think these changes should be part of your completion PR.

File: .vscode/launch.json (hunk 5 above)
Comment (dobladez) [reply .vscode/launch.json:620]: No mistake. It was duplicated.

File: packages/salesforcedx-vscode-soql/src/client/client.ts (hunk 6 above)
Comment (dobladez) [reply packages/salesforcedx-vscode-soql/src/client/client.ts:84]: It is... I've separated the completion vs. query validation into different files. `client.ts` was getting too big and with many responsibilities (initialization, completion, query validation)

File: packages/salesforcedx-vscode-soql/src/client/codeCompletion.ts (hunk 7 above)
Comment (dobladez) [reply packages/salesforcedx-vscode-soql/src/client/codeCompletion.ts:102]: It's just code moved from client.ts into a separate file. See commit dc95f508c5f116b5b39e3d08251b62ab4a8a33cb

**Review by praksb — APPROVED** [review]

**Review by jag-sfdc — APPROVED** [review]
